src/main/java/io/github/jancieslak/tba/services/ReservationService.java

```diff
@@ -0,0 +1,151 @@
+package io.github.jancieslak.tba.services;
+
+import io.github.jancieslak.tba.dto.PostReserveScreeningRequestModel;
+import io.github.jancieslak.tba.dto.PostReserveScreeningResponseModel;
+import io.github.jancieslak.tba.dto.SeatReservationModel;
+import io.github.jancieslak.tba.model.Reservation;
+import io.github.jancieslak.tba.model.ReservedSeat;
+import io.github.jancieslak.tba.model.Screening;
+import io.github.jancieslak.tba.model.Ticket;
+import io.github.jancieslak.tba.repository.*;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@Service
+public class ReservationService {
+    private final static Pattern FIRSTNAME_PATTERN = Pattern.compile("([A-ZĄĆĘŁŃÓŚŹŻ][a-ząćęłńóśźż]{2,})");
+    private final static Pattern LASTNAME_PATTERN = Pattern.compile("([A-ZĄĆĘŁŃÓŚŹŻ][a-ząćęłńóśźż]{2,})(/[A-ZĄĆĘŁŃÓŚŹŻ][a-ząćęłńóśźż]{2,})?");
+
+    private final ReservedSeatsRepository reservedSeatsRepository;
+    private final ScreeningRepository screeningRepository;
+    private final ReservationRepository reservationRepository;
+    private final ScreeningRoomRepository screeningRoomRepository;
+
+    public ReservationService(ReservedSeatsRepository reservedSeatsRepository,
+                              ScreeningRepository screeningRepository,
+                              ReservationRepository reservationRepository,
+                              ScreeningRoomRepository screeningRoomRepository) {
+        this.reservedSeatsRepository = reservedSeatsRepository;
+        this.screeningRepository = screeningRepository;
+        this.reservationRepository = reservationRepository;
+        this.screeningRoomRepository = screeningRoomRepository;
+    }
+
+    // Check for single place between already reserved seats
+    // If there is such place the seats placement is not valid otherwise it's valid
+    public boolean hasValidSeatsPlacement(Set<SeatReservationModel> seats, Set<ReservedSeat> reservedSeats, Screening screening) {
+        // Get request seats columns grouped by row
+        var partitionedRequestSeats = seats.stream()
+                .collect(Collectors.groupingBy(SeatReservationModel::getRow));
+
+        // Get reserved seats columns filtered by request seats rows
+        var partitionedReservedSeats = reservedSeats.stream()
+                .filter(reservedSeat -> partitionedRequestSeats.get(reservedSeat.getSeat().getRow().getScreeningRoomRow()) != null)
+                .collect(Collectors.groupingBy(reservedSeat -> reservedSeat.getSeat().getRow().getScreeningRoomRow()));
+
+        var screeningRoomOption = screeningRoomRepository.findScreeningRoomByScreeningId(screening.getId());
+
+        if (screeningRoomOption.isEmpty()) {
```

Comment:
wołanie isEmpty/isPresent a potem get to trochę smrodek w kodzie

Reply:
Próbowałem na wiele sposobów na ładne wyjęcie wartości z Optional-a i w tej sytuacji chyba tylko .orElseThrow wydawałby się najlepszy tylko nie wiedziałem czy w tej sytuacji wyrzucanie wyjątków jest ok